File: 008/Erlang/problem008.erl
-module(problem008).
-export([largest_five_digit_product/1, largest_five_digit_product/2]).

-import(lists, [max/1]).

largest_five_digit_product(N) -> largest_five_digit_product(list_of_digits(N), 0).

largest_five_digit_product(L, Current) when length(L) < 5 -> Current;
largest_five_digit_product([H|T], Current) ->
  largest_five_digit_product(T, max([Current, product5([H|T])])).

list_of_digits(0) -> [];
list_of_digits(N) when N < 10 -> [N];
list_of_digits(N) -> list_of_digits(N div 10) ++ [N rem 10].

product5([A, B, C, D, E | _]) -> A * B * C * D * E;
product5(_) -> 0.

File: 008/Haskell/Problem008.hs
module Problem008 (largestFiveDigitProduct) where

largestFiveDigitProduct :: Integer -> Integer
largestFiveDigitProduct n = lfdp (listify n []) 0
  where
    lfdp list@(a:b:c:d:e:rest) current = lfdp (tail list) (max current (a*b*c*d*e))
    lfdp _ current = current
    listify 0 list = list
    listify n list = listify (n `div` 10) (n `mod` 10 : list)
